lib/dashboard/adapter.go

```diff
@@ -0,0 +1,193 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dashboard
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"sort"
+	"strings"
+	"time"
+
+	// nolint[lll]
+	containeranalysis "cloud.google.com/go/containeranalysis/apiv1"
+	"cloud.google.com/go/storage"
+	"google.golang.org/api/iterator"
+	grafeaspb "google.golang.org/genproto/googleapis/grafeas/v1"
+)
+
+// nolint[errcheck]
+func uploadFile(object, bucket string) error {
+	const timeout = 60
+	ctx := context.Background()
+	client, err := storage.NewClient(ctx)
+	if err != nil {
+		return fmt.Errorf("storage.NewClient: %v", err)
+	}
+	defer client.Close()
+
+	// Open local file.
+	f, err := os.Open("/Users/yodalemu/google/myfork/k8s-container-image-promoter/" + object)
+	if err != nil {
+		return fmt.Errorf("os.Open: %v", err)
+	}
+	defer f.Close()
+
+	ctx, cancel := context.WithTimeout(ctx, time.Second*timeout)
+	defer cancel()
+
+	// Upload an object with storage.Writer.
+	wc := client.Bucket(bucket).Object(object).NewWriter(ctx)
+	if _, err = io.Copy(wc, f); err != nil {
+		return fmt.Errorf("io.Copy: %v", err)
+	}
+	if err := wc.Close(); err != nil {
+		return fmt.Errorf("Writer.Close: %v", err)
+	}
+	return nil
+}
+
+// GetAllVulnerabilities gets all of the vulnerability occurrences associated
+// with images in a specific project using the Container Analysis Service.
+// nolint[errcheck]
+func GetAllVulnerabilities(
+	projectID string,
+) ([]*grafeaspb.Occurrence, error) {
+	ctx := context.Background()
+	client, err := containeranalysis.NewClient(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("NewClient: %v", err)
+	}
+	defer client.Close()
+
+	req := &grafeaspb.ListOccurrencesRequest{
+		Parent: fmt.Sprintf("projects/%s", projectID),
+		Filter: fmt.Sprintf("kind = %q", "VULNERABILITY"),
+	}
+
+	var occurrenceList []*grafeaspb.Occurrence
+	it := client.GetGrafeasClient().ListOccurrences(ctx, req)
+	for {
+		occ, err := it.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return nil, fmt.Errorf("occurrence iteration error: %v", err)
+		}
+		occurrenceList = append(occurrenceList, occ)
+	}
+	return occurrenceList, err
+}
+
+func parseImageResourceURL(resourceURL string) (string, string) {
+	FQIN := path.Base(resourceURL)
+	splitFQIN := strings.Split(FQIN, "@")
+	return splitFQIN[0], splitFQIN[1]
+}
+
+func parseVulnName(noteName string) string {
+	return path.Base(noteName)
+}
+
+// GenerateDashboardJSON parses the a slice of vulnerability occurrences
+// into a JSON that only contains the necessary information for each
+// vulnerability. The JSON is also formatted to be easily processed into
+// an HTML table by the vulnerability dashboard.
+func GenerateDashboardJSON(
+	productionVulnerabilities []*grafeaspb.Occurrence,
+) ([]byte, error) {
+	var occByResource = make(map[string]ImageVulnOcc)
+	var resourceURIs []string
```

Comment:
Please explain what a resourceURI is, and give an actual example of what one might look like.

Reply:
Done